### DISCUSSION TITLE
Parse the incoming byteArr from receive example

### DISCUSSION THREAD

**Discussion by TinManAkshay** [jgromes/RadioLib#214] [Q&A]
Hi,

So far I have been receiving the correct data using Receiver example. Moreover, I need to parse the uint8_t* byteArr data also. Till now, I have tried lot of methods to do that, but i was unsuccessful. I have been focusing more on converting the data into char array, then using strtok() technique to parse it, but not printing anything ( no compilation error though). I wonder if its right way to parse the incoming uint8_t data. 

If anyone has already tried parsing the data received from .receive() function or any idea, please do suggest here. 

Thanks a lot,
Akshay

**Comment by jgromes** [comment]
I'm not sure what you'Re trying to achieve - `strtok()` breaks a C-string (a null-terminated char array) by some token, e.g. you can break up a sentence into words by the space character.

I guess you should start by defining what exactly you mean by "parsing". Are you trying to print the binary data as hexadecimal numbers? And what's the data you're working with?

EDIT: Also, not an issue - converting to discussion.

**Comment by TinManAkshay** [comment]
For an instance,

I am transmitting the command "$setconf,106" from the SX1272 transmitter to the SX1262 receiver and receiver receives it using radio.receive(byteArr, 32). I am printing the command using for loop. Then I converted the received data to char array so that I can use strtok() to parse $setconf and 106 and store them in a different variables. See "," is a delimiter here.

Also, in the received data, rest of the data is just garbage. I want to get rid of it too once we parse two parameters from the command. 

Hope you got my point now.

**Reply by jgromes** [reply]
If your command is always a string, why do you need to use a byte array? If you call radio.receive(byteArr, 32), then you force the radio to always try to read 32 bytes, regardless of what was received. There won't be a terminating null byte at the end, hence the garbage at the end.

I think you're overthinking this: send a string with radio.transmit("$setconf,106"); then receive it with radio.receive(str); and process the str by converting that into char array.

**Comment by TinManAkshay** [comment]
Its a string in the form of bytes. See my transmitter is built on SX172 python libraries. There I am sending the command $setconf,106 in the form of bytes to SX1262. The whole code is made that way. So radio.receive(str) did not work here. So I believe I have to continue using radio.receive(byteArr, len) in order to receive the commands. In this case, how do you suggest to parse the byteArr data??

**Reply by jgromes** [reply]
As far as I know, literally every data type is "in the form of bytes". Have you tried receiving as string? And if so, what was the result? In transmission, there's no difference between "string" or "bytes" or any other datatype, it's all just ... well, bytes.

**Comment by TinManAkshay** [comment]
Yes I tried. It was all blank. When I use the other way like receive(byteArr), I can see the whole command on the serial monitor. I just need to parse the parameters separated by commas and store somewhere for further analysis.

**Reply by jgromes** [reply]
Then you should check what exactly are you sending - if it's for example "\0$setconf,106\0" (\0 being null bytes), then the string will be terminated at the first character. There's no fundamental difference between a char array and a byte array - it's all just bytes in a memory. The only difference between a byte array and a string is that by convention, strings in C are null-terminated.

**Reply by tauesdospi** [reply]
Hi sorry for wake up this topic but I'm having issues with transmit/receive when using string (it always works) and when using byte array (because I'm not sending ASCII text but binary message variable length). In this last case, when receving ( readData(buffer, 0) because it has variable length) sometimes I get the correct data, sometimes previous message data is copied in part in the new receive, also after 2nd o 3rd readData, the packet length is inserted in the first byte of the message.

Is there any issue with binary data being send/received? (because in the examples you are always using ascii strings)
Thanks
Antonio

**Reply by jgromes** [reply]
> in the examples you are always using ascii strings

... which internally get converted to byte arrays. So my guess is there is something wrong in your sketch (which you have not posted).

**Reply by tauesdospi** [reply]
Hi my sketch is very easy.. this is a really weird thing..

basically:

SX1276 radio = new Module(GPIO_NUM_18, GPIO_NUM_26, GPIO_NUM_14, GPIO_NUM_35);

void setup()
{
....
 state = radio.beginFSK(); //868.8, 50, 25, 125, 17, 5, false);    
     state = radio.setFrequency(868.8);
     state = radio.setBitRate(50.0);
     state = radio.setFrequencyDeviation(25);
     state = radio.setRxBandwidth(50.0);
     state = radio.setOutputPower(17.0);
     state = radio.setCurrentLimit(100);
     state = radio.setAFCBandwidth(83.3);
     state = radio.setDataShaping(RADIOLIB_SHAPING_1_0);
     uint8_t syncWord[] = {0xC1, 0x94, 0xC1 };
     state = radio.setSyncWord(syncWord, 3);
     state = radio.setPreambleLength(40);
     state = radio.disableAddressFiltering();
     state = radio.setEncoding(RADIOLIB_ENCODING_WHITENING);
     radio.setDio0Action(OnFSKCompleted, RISING);
     state = radio.setCRC(true);
     state = radio.variablePacketLengthMode();
}


volatile bool pkt_done = false;
void OnFSKCompleted(void)
{
  pkt_done = true;
}

void loop(void)
{
static estado = 98;   // state machine starts at state 98
uint8_t data[63];
...
switch(estado)
{
case 98:
     pkt_done = false;
     data[0] = 'p';
     data[1] = millis() & 0xff;
     data[2] = 'n';
     data[3] = 'g';
     state = radio.startTransmit(data,4);
     estado = 99;
     break;

   case 99:
     if(pkt_done)
     { 
       pkt_done = false;
       state = radio.startReceive();                              // esperamos respuesta
       estado = 100;
     }
     break;

    case 100:
     if(pkt_done)
     {
       state = radio.readData(data,0);                           // leemos el pkt en el buffer
       pkt_len = radio.getPacketLength();
       Serial.printf("rx pkt len=%d -- ", pkt_len);
       for(uint8_t n = 0; n < pkt_len; n++) { Serial.printf("%02x ", data[n]); } Serial.println("");
       estado = 98;
     }
     break;
}
...
vTaskDelay(1 ms)
}

basically this code transmit a 4 byte message; it is always correctly received by the peer, then remains pending of receiving a variable length packet from the peer.

If I send packet "12345" it shows:
rx pkt len=6 -- 31 32 33 34 35 0a  ==> watch the 0x0a at the end ! and getPacketLength() is returning 6, not 5 !
if I send again the packet "12345" it shows:
rx pkt len=6 -- 06 31 32 33 34 35 ==> now the 1st byte is 0x06 ! the incorrect packet length ?
if I send again "12345" :
rx pkt len=6 -- 06 31 32 33 34 35  => incorrect again
if I send now "1234567890123456789012345678901234567890" then I receive:
23:05:55.637 -> rx pkt len=41 -- 29 31 32 33 34 35 31 34 30 34 0a 00 00 00 00 00 00 00 00 0e 4b 0d 80 e0 27 fb 3f 02 00 00 00 00 00 00 00 dc 2a fc 3f 23 06  ==> absolutely incorrect packet content !! only 1st 5 bytes are correct (remember previous packets were 5 byte len)
if I send again 1234567890123456789012345678901234567890 then I receive now:
rx pkt len=41 -- 29 31 32 33 34 35 36 37 38 39 30 31 32 33 34 35 36 37 38 39 30 31 32 33 34 35 36 37 38 39 30 31 32 33 34 35 36 37 38 39 30   ==> 1st byte is length 0x29 = 41 which is also incorrect, should be 40

If now I get the exactly same code and replace the  byte array data by String and update the readData(data, 0) => readData(string) and send only ascii characters (because 0x0 is considered end of string!), then everything works perfectly.

I'm using the Heltec module "Wireless Stick Lite" which has a SX1276 lora/fsk module.

Hope you have a chance to see my code and let me what is wrong!
Thanks
Antonio

**Reply by tauesdospi** [reply]
Hi,

just replacing in the above code:

       //state = radio.readData(data,0);                           // leemos el pkt en el buffer
       //pkt_len = radio.getPacketLength();
       state = radio.readData(s);
       pkt_len = s.length();
       s.getBytes(data, pkt_len);

the readData(char *, len)  is replaced by readData(string&)

and the code works perfectly... all packets received correctly no the packet len copied at the beggining or the end of the packet, no invalid packet content the 1st time the packet is send ...

It seems the the readData(char *,len) function used directly by me in the code is not the same function readData(string&) is using internally to convert the char * to string??

Can you please advise?

Thanks
Antonio

**Reply by jgromes** [reply]
@tauesdospi your order of operations is quite strange:

```c++
state = radio.readData(data,0);
pkt_len = radio.getPacketLength();
```

First you try to read the data and only then attempt to figure out how much was received? Why is that?

**Reply by tauesdospi** [reply]
Hi Jan,
absolutely clear right now! Just inverting the lines everything is working like a charm. I misunderstand the docs.

readLen:
"Reads data that was received after calling startReceive method" and "Number of bytes that will be read. When set to 0, the packet length will be retreived automatically."

As I don't know the received packet length in advance I guessed readLen(buffer, 0), then getPacketLength for determining the length.

Best regards and thank you for your support !
Antonio

**Reply by jgromes** [reply]
@tauesdospi it is true that neither the examples, nor the API reference mentions that this order of operations must be followed. I added these notes, thanks for pointing this out.
